File: go/dp_one_dim/322_coin_change.go
package dponedim

import "math"

// coinChange 零钱兑换
func coinChange(coins []int, amount int) int {
	// coins [1,2,5] aoumt 11
	// dp[11], 取1 dp[10] + 1
	// dp[11], 取2 dp[9] + 1
	// dp[11], 取5 dp[6] + 1
	// dp[11] = min(dp[10], dp[9], dp[6]) +1
	// dp[n] = min(amount - dp[i]) + 1

	dp := make([]int, amount+1)
	// 初始化位最大值
	for i := 0; i < amount+1; i++ {
		dp[i] = math.MaxInt32
	}
	dp[0] = 0
	// 穷举每一个目标数的最小硬币数
	for i := 1; i <= amount; i++ {
		//对每种子问题都穷举
		for _, v := range coins {
			if v <= i {
				dp[i] = min(dp[i], dp[i-v]+1)
			}
		}
	}

	if dp[amount] == math.MaxInt32 {
		return -1
	}

	return dp[amount]
}

File: rust/src/dp_one_dim/_322_coin_change.rs
/// 零钱兑换
pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
    // 最大的硬币数是amout，因为硬币的面值最小是1
    let max_val = amount + 1;

    let amount = amount as usize;
    let mut dp = vec![max_val; amount + 1];
    dp[0] = 0;

    // 穷举每个目标数的最小硬币数
    for i in 1..=amount {
        //对每种子问题都穷举
        for v in coins.iter() {
            // v > i 的时候已经无法满足
            if *v as usize <= i {
                dp[i] = dp[i].min(dp[i - *v as usize] + 1);
            }
        }
    }

    if dp[amount] == max_val {
        return -1;
    }
    dp[amount]
}
